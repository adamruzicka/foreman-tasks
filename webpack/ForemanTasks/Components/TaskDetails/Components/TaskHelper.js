import { translate as __, documentLocale } from 'foremanReact/common/I18n';
import React from 'react';
import { FormattedRelative } from 'react-intl';
import humanizeDuration from 'humanize-duration';

const formatDate = date => {
  // Firefox doesnt format dd-mm-yyyy type strings to date
  if (typeof date === 'string' || date instanceof String)
    return new Date(date.replace(/-/, '/').replace(/-/, '/'));
  return date;
};

export const timeInWords = time => {
  if (!time) return __('N/A');
  time = formatDate(time);
  return <FormattedRelative value={time} />;
};

export const durationInWords = (
  start,
  finish,
  selectedLocale = documentLocale()
) => {
  if (!start) return __('N/A');
  start = formatDate(start).getTime();
  finish = formatDate(finish).getTime();
  return {
    text: humanizeDuration(new Date(finish - start).getTime(), {
      largest: 1,
      language: selectedLocale,
      fallbacks: ['en'],
    }),
    tooltip: `${numberWithDelimiter((finish - start) / 1000)} ${__('seconds')}`,
  };
};

const numberWithDelimiter = x =>
  x.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');
