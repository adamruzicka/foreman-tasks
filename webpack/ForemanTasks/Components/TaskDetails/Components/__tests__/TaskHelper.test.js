import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { IntlProvider } from 'react-intl';
import { timeInWords, durationInWords } from '../TaskHelper';

describe('timeInWords', () => {
  it('should work for past minutes', () => {
    const component = mount(
      <IntlProvider locale="en">
        {timeInWords(new Date().getTime() - 1000 * 60 * 5)}
      </IntlProvider>
    );
    expect(toJson(component.render())).toMatchSnapshot();
  });
  it('should work for past hours', () => {
    const component = mount(
      <IntlProvider locale="en">
        {timeInWords(new Date().getTime() - 1000 * 60 * 60 * 5)}
      </IntlProvider>
    );
    expect(toJson(component.render())).toMatchSnapshot();
  });
  it('should work for past days', () => {
    const component = mount(
      <IntlProvider locale="en">
        {timeInWords(new Date().getTime() - 1000 * 60 * 60 * 24 * 5)}
      </IntlProvider>
    );
    expect(toJson(component.render())).toMatchSnapshot();
  });
  it('should work for future minutes', () => {
    const component = mount(
      <IntlProvider locale="en">
        {timeInWords(new Date().getTime() + 1000 * 60 * 5)}
      </IntlProvider>
    );
    expect(toJson(component.render())).toMatchSnapshot();
  });
  it('should work for future hours', () => {
    const component = mount(
      <IntlProvider locale="en">
        {timeInWords(new Date().getTime() + 1000 * 60 * 60 * 5)}
      </IntlProvider>
    );
    expect(toJson(component.render())).toMatchSnapshot();
  });
  it('should work for future days', () => {
    const component = mount(
      <IntlProvider locale="en">
        {timeInWords(new Date().getTime() + 1000 * 60 * 60 * 24 * 5)}
      </IntlProvider>
    );
    expect(toJson(component.render())).toMatchSnapshot();
  });
});

describe('durationInWords', () => {
  it('should work for seconds', () => {
    expect(durationInWords('1/1/1 10:00:00', '1/1/1 10:00:01')).toEqual({
      text: '1 second',
      tooltip: '1 seconds',
    });
  });
  it('should work for minutes', () => {
    expect(durationInWords('1/1/1 10:00:00', '1/1/1 10:02:01')).toEqual({
      text: '2 minutes',
      tooltip: '121 seconds',
    });
  });
  it('should work for hours', () => {
    expect(durationInWords('1/1/1 10:00:00', '1/1/1 13:00:01')).toEqual({
      text: '3 hours',
      tooltip: '10,801 seconds',
    });
  });
});
