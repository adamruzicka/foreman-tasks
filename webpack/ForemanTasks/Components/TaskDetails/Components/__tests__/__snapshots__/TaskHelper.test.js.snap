// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`timeInWords should work for future days 1`] = `
<span>
  in 5 days
</span>
`;

exports[`timeInWords should work for future hours 1`] = `
<span>
  in 5 hours
</span>
`;

exports[`timeInWords should work for future minutes 1`] = `
<span>
  in 5 minutes
</span>
`;

exports[`timeInWords should work for past days 1`] = `
<span>
  5 days ago
</span>
`;

exports[`timeInWords should work for past hours 1`] = `
<span>
  5 hours ago
</span>
`;

exports[`timeInWords should work for past minutes 1`] = `
<span>
  5 minutes ago
</span>
`;
